src/assignments/phat/lesson_3_answers.clj

```diff
@@ -0,0 +1,143 @@
+              :rank (card-rank (last (str/split card #""))))))
+
+;; tests
+(= {:suit :diamond :rank 10} (recognize-card "DQ"))
+(= {:suit :heart :rank 3} (recognize-card "H5"))
+(= {:suit :club :rank 12} (recognize-card "CA"))
+(= (range 13) (map (comp :rank recognize-card str)
+                   '[S2 S3 S4 S5 S6 S7
+                     S8 S9 ST SJ SQ SK SA]))
+
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; 178 - Best hand
+
+(defn sort-cards
+  "Recognize all cards and group cards by suit or rank"
+  [card-coll keyword]
+  (into (sorted-map) (group-by keyword (map recognize-card card-coll))))
+
+
+(defn continuous-sequence?
+  "check if a sequence increases with a step of 1"
+  [coll]
+  (apply = 1 (map - (rest coll) coll)))
+
+
+(defn count-ranks
+  "return a sequence of count of each card rank"
+  [card-coll]
+  (for [x (vals (sort-cards card-coll :rank))]
+    (count x)))
+
+
+(defn straight-flush?
+  "All cards in the same suit, and in sequence"
+  [card-coll]
+  (and (= 1 (count (keys (sort-cards card-coll :suit))))
+       (continuous-sequence? 
+        (sort < (keys (sort-cards card-coll :rank))))))
+
+
+(defn four-of-a-kind?
+  "Four of the cards have the same rank"
+  [card-coll]
+  (if (some #(= 4 %) (count-ranks card-coll))
+    true
+    false))
+
+
+(defn full-house?
+  "Three cards of one rank, the other two of another rank"
+  [card-coll]
+  (let [counts (count-ranks card-coll)]
+    (if (and (some #(= 3 %) counts)
+             (some #(= 2 %) counts))
+      true
+      false)))
+
+
+(defn flush?
+  "All cards in the same suit"
+  [card-coll]
+  (= 1 (count (keys (sort-cards card-coll :suit)))))
+
+
+(defn straight?
+  "All cards in sequence (aces can be high or low, but not both at once)"
+  [card-coll]
+  (let [seq (sort < (keys (sort-cards card-coll :rank)))]
+    (if (= '(0 1 2 3 12) seq)
+      true
+      (continuous-sequence? seq))))
+
+
+(defn three-of-a-kind?
+  "Three of the cards have the same rank"
+  [card-coll]
+  (if (some #(= 3 %) (count-ranks card-coll))
+    true
+    false))
+
+
+(defn two-pair?
+  "Two pairs of cards have the same rank"
+  [card-coll]
+  (= 2 ((frequencies (count-ranks card-coll)) 2)))
+
+
+(defn pair?
+  "Two cards have the same rank"
+  [card-coll]
+  (= 1 ((frequencies (count-ranks card-coll)) 2)))
+
+
+(defn best-hand
+  "take a collection of 5 cards  and determine the best poker hand"
+  [card-coll]
+  (cond
+    (straight-flush? card-coll) :straight-flush
+    (four-of-a-kind? card-coll) :four-of-a-kind
+    (full-house? card-coll) :full-house
+    (flush? card-coll) :flush
+    (straight? card-coll) :straight
+    (three-of-a-kind? card-coll) :three-of-a-kind
+    (two-pair? card-coll) :two-pair
+    (pair? card-coll) :pair
+    :else :high-card))
+
+
+;; tests
+(= :high-card (best-hand ["HA" "D2" "H3" "C9" "DJ"]))
+(= :pair (best-hand ["HA" "HQ" "SJ" "DA" "HT"]))
+(= :two-pair (best-hand ["HA" "DA" "HQ" "SQ" "HT"]))
```

Comment:
Very nice.  Did you get a chance to try to build a test harness with deftest?

Reply:
Sorry, I didn't. I will try to add a deftest after the lesson tonight.